### ISSUE TITLE
`GradientStop.Offset` Bindings Don't Work

### ISSUE THREAD

**Issue by colejohnson66** [AvaloniaUI/Avalonia#8266]
**Describe the bug**
If you put a binding on the `Offset` property of `GradientStop`, you'd expect that updating the value of the binding would cause a redraw, but it doesn't.

**To Reproduce**

1. `<GradientStop Color="Black" Offset="{Binding ...}" />`
2. Update binding value
3. Observe no effect on the gradient
4. Force a redraw by doing something to a parent control
5. Observe gradient changing

**Expected behavior**
Updating a binding works

**Desktop (please complete the following information):**

- OS: Windows
- Version 0.10.14

**Comment by colejohnson66** [comment]
https://user-images.githubusercontent.com/11381599/171905990-801baa7a-03ba-4b05-bac3-5c666674235c.mp4

```cs
public class OpacityFadeVisibilityConverter : IMultiValueConverter
{
    private static readonly double FADE_CHECK = 50;
    public static readonly OpacityFadeVisibilityConverter Instance = new();

    public object Convert(IList<object?> values, Type targetType, object? parameter, CultureInfo culture)
    {
        if (parameter is not string param || values.Count != 3)
            return AvaloniaProperty.UnsetValue;

        if (values[0] is not double offset || values[1] is not double extent || values[2] is not double viewport)
            return AvaloniaProperty.UnsetValue;

        // return a double between 0.0 and 1.0 that is FADE_CHECK pixels from 0.0 or 1.0
        // shift the returned value closer to 0.0/1.0 if less than FADE_CHECK pixels are offset

        double fadeCheckAsPercent = FADE_CHECK / viewport;

        if (param == "0")
        {
            if (offset < FADE_CHECK)
                return ScaleIntoRange(offset, (0, FADE_CHECK), (0, fadeCheckAsPercent));
            return fadeCheckAsPercent;
        }

        Debug.Assert(param is "100");
        double endCheck = Math.Abs(extent - viewport - offset);
        if (endCheck < FADE_CHECK)
            return ScaleIntoRange(FADE_CHECK - endCheck, (0, FADE_CHECK), (1.0 - fadeCheckAsPercent, 1.0));
        return 1.0 - fadeCheckAsPercent;
    }

    private static double ScaleIntoRange(double value, (double, double) from, (double, double) into)
    {
        double scale = (into.Item2 - into.Item1) / (from.Item2 - from.Item1);
        return (value - from.Item1) * scale + into.Item1;
    }
}
```

```axaml
    <Style Selector="ScrollViewer.OpacityFadeX">
        <Setter Property="Background" Value="Transparent" />
        <Setter Property="Template">
            <ControlTemplate>
                <Grid
                    ColumnDefinitions="*,Auto"
                    RowDefinitions="*,Auto">

                    <Panel
                        Grid.Row="0" Grid.Column="0"
                        Grid.RowSpan="2">
                        <Panel.OpacityMask>
                            <LinearGradientBrush StartPoint="0%,50%" EndPoint="100%,50%">
                                <LinearGradientBrush.GradientStops>
                                    <GradientStop Offset="0" Color="Transparent" />
                                    <GradientStop Color="Black">
                                        <GradientStop.Offset>
                                            <MultiBinding
                                                Converter="{x:Static conv:OpacityFadeVisibilityConverter.Instance}"
                                                ConverterParameter="0">
                                                <Binding RelativeSource="{RelativeSource TemplatedParent}" Path="Offset.X" />
                                                <Binding RelativeSource="{RelativeSource TemplatedParent}" Path="Extent.Width" />
                                                <Binding RelativeSource="{RelativeSource TemplatedParent}" Path="Viewport.Width" />
                                            </MultiBinding>
                                        </GradientStop.Offset>
                                    </GradientStop>
                                    <GradientStop Color="Black">
                                        <GradientStop.Offset>
                                            <MultiBinding
                                                Converter="{x:Static conv:OpacityFadeVisibilityConverter.Instance}"
                                                ConverterParameter="100">
                                                <Binding RelativeSource="{RelativeSource TemplatedParent}" Path="Offset.X" />
                                                <Binding RelativeSource="{RelativeSource TemplatedParent}" Path="Extent.Width" />
                                                <Binding RelativeSource="{RelativeSource TemplatedParent}" Path="Viewport.Width" />
                                            </MultiBinding>
                                        </GradientStop.Offset>
                                    </GradientStop>
                                    <GradientStop Offset="1" Color="Transparent" />
                                </LinearGradientBrush.GradientStops>
                            </LinearGradientBrush>
                        </Panel.OpacityMask>

                        <!-- actual content -->
                        <ScrollContentPresenter
                            x:Name="PART_ContentPresenter"
                            Background="{TemplateBinding Background}"
                            CanHorizontallyScroll="{TemplateBinding CanHorizontallyScroll}"
                            CanVerticallyScroll="{TemplateBinding CanVerticallyScroll}"
                            Content="{TemplateBinding Content}"
                            Extent="{TemplateBinding Extent, Mode=TwoWay}"
                            Margin="{TemplateBinding Padding}"
                            Offset="{TemplateBinding Offset, Mode=TwoWay}"
                            Viewport="{TemplateBinding Viewport, Mode=TwoWay}">
                            <ScrollContentPresenter.GestureRecognizers>
                                <ScrollGestureRecognizer
                                    CanHorizontallyScroll="{TemplateBinding CanHorizontallyScroll}"
                                    CanVerticallyScroll="{TemplateBinding CanVerticallyScroll}" />
                            </ScrollContentPresenter.GestureRecognizers>
                        </ScrollContentPresenter>
                    </Panel>

                    <!-- bottom (horizontal) scroll bar -->
                    <ScrollBar
                        x:Name="PART_HorizontalScrollBar"
                        Grid.Row="1" Grid.Column="0"
                        AllowAutoHide="False"
                        Focusable="False"
                        LargeChange="{Binding LargeChange.Width, RelativeSource={RelativeSource TemplatedParent}}"
                        Maximum="{TemplateBinding HorizontalScrollBarMaximum}"
                        Orientation="Horizontal"
                        SmallChange="{Binding SmallChange.Width, RelativeSource={RelativeSource TemplatedParent}}"
                        Value="{TemplateBinding HorizontalScrollBarValue, Mode=TwoWay}"
                        ViewportSize="{TemplateBinding HorizontalScrollBarViewportSize}"
                        Visibility="{TemplateBinding HorizontalScrollBarVisibility}" />

                    <!-- right (vertical) scroll bar -->
                    <ScrollBar
                        x:Name="PART_VerticalScrollBar"
                        Grid.Row="0" Grid.Column="1"
                        AllowAutoHide="{TemplateBinding AllowAutoHide}"
                        Focusable="False"
                        LargeChange="{Binding LargeChange.Height, RelativeSource={RelativeSource TemplatedParent}}"
                        Maximum="{TemplateBinding VerticalScrollBarMaximum}"
                        Orientation="Vertical"
                        SmallChange="{Binding SmallChange.Height, RelativeSource={RelativeSource TemplatedParent}}"
                        Value="{TemplateBinding VerticalScrollBarValue, Mode=TwoWay}"
                        ViewportSize="{TemplateBinding VerticalScrollBarViewportSize}"
                        Visibility="{TemplateBinding VerticalScrollBarVisibility}"
                        ZIndex="2" />

                    <Panel
                        x:Name="PART_ScrollBarsSeparator"
                        Grid.Row="1" Grid.Column="1"
                        Background="{DynamicResource ScrollViewerScrollBarsSeparatorBackground}" />
                </Grid>
            </ControlTemplate>
        </Setter>
    </Style>
````

**Comment by colejohnson66** [comment]
Every time the `<ScrollViewer>` moves, the converter is being called (breakpoints are hit). The returned value is just ignored until the outer `<ScrollViewer>` scrolls

**Comment by timunie** [comment]
Can you test the same brush without using it as opacity mask? I wonder if it's the brush that doesn't update or the parent that does not handle changes in opacity mask

**Comment by colejohnson66** [comment]
Yep; It's because of the opacity mask. Changing it to `<Panel.Background>` (and the transparent stops to white) works fine.

https://user-images.githubusercontent.com/11381599/171933387-de5180df-9316-4520-a160-7a596a49dd9c.mp4

**Comment by colejohnson66** [comment]
It gets weirder: It works in the XAML previewer in Rider, just **not** in our theme demo.

**Comment by maxkatz6** [comment]
Designer uses immediate rendering mode iirc.

**Comment by maxkatz6** [comment]
Try to execute this code once in your app, it should trigger invalidation for deferred renderer:
```
GradientStop.OffsetProperty.Changed.Subscribe(e => ((GradientStop)e).RaiseInvalidated(EventArgs.Empty));
```

**Comment by colejohnson66** [comment]
Is that a 0.11 API? `e` is of type `AvaloniaPropertyChangedEventArgs<double>` and `(GradientStop)e.Sender` doesn't have a `RaiseInvalidated` function.

**Comment by maxkatz6** [comment]
No, I was wrong. GradientStop doesn't implement IAffectsRender at all.

**Comment by maxkatz6** [comment]
Linear gradient brush needs to subscribe on children updates (offset and brush) and call RaiseInvalidated inside it. 
To unblock yourself it should be possible to inherit LinearGradientBrush and subscribe on gradient stops changes in your custom brush type.

**Comment by colejohnson66** [comment]
Well, `LinearGradientBrush` is sealed... So I had to make a dumb copy of Avalonia's with this single change:

```cs
    public LinearGradientBrush2()
    {
        GradientStop.OffsetProperty.Changed.Subscribe(e => RaiseInvalidated(e));
    }
```

And it didn't work.

**Comment by colejohnson66** [comment]
I am running it against the brush

**Comment by maxkatz6** [comment]
@colejohnson66 RaiseInvalidated gets called with GradientStop as an parameter, not the brush

**Comment by colejohnson66** [comment]
I'm not sure what you mean. I'm calling `RaiseInvalidated` against my own `LinearGradientBrush2` object, whether I do `this.RaiseInvalidated` or just `RaiseInvalidated`. If you meant passing `e` into the function, changing it to this:

```cs
public LinearGradientBrush2()
{
    GradientStop.OffsetProperty.Changed.Subscribe(_ => RaiseInvalidated(EventArgs.Empty));
}
```

...did not fix the issue.

**Comment by colejohnson66** [comment]
Well, 11.0-preview1 seems to have made it worse... Now, scrolling the parent control vertically doesn't seem to trigger a redraw (like it did in my video).

**Comment by colejohnson66** [comment]
Still broken on 11.0-preview3.

**Comment by colejohnson66** [comment]
I found this "neat" piece of code that appears to do the actual subscribing to the `Invalidated` event for "visual" objects:

https://github.com/AvaloniaUI/Avalonia/blob/a344fecc9534ea69924fe67d08c3202959e545b5/src/Avalonia.Base/Visual.cs#L104-L107

That weak event is used inside `AffectsRender`:

https://github.com/AvaloniaUI/Avalonia/blob/a344fecc9534ea69924fe67d08c3202959e545b5/src/Avalonia.Base/Visual.cs#L375-L386

*However*, this code only ever gets used if I write `<Panel.Background>`, not `<Panel.OpacityMask>`. With `OpacityMask`, a conditional breakpoint on line 105 above (`s is LinearGradientBrush2`) never triggers, but `Background` does.

**Comment by timunie** [comment]
@colejohnson66 I guess you need to copy GradientBrushs code and implement IBrush instead. We made things internal where we have changes planned that would be a breaking change as we cannot (don't want to at least) make breaking changes in the 11.x timeframe.

**Comment by colejohnson66** [comment]
I figured I'd check back now, as it's still broken on 11.0.10, and I can't even implement `IBrush` as it's annotated with `NotClientImplementable`. Trying to just inherit from `Brush` directly right now...

![image](https://github.com/AvaloniaUI/Avalonia/assets/11381599/ec422761-0f5c-421a-b0d5-4fef87c2e470)

**Comment by theoractice** [comment]
I can confirm that this is still an issue for 11.3.8.
